### PR TITLE
[CR] Centralize book readability checks

### PR DESCRIPTION
**Pull request by sacrificerXY**
#### Summary
SUMMARY: Infrastructure "Centralize book readability checks"

#### Purpose of change

I wanted to refactor the book reading routines to use the new `activity_actor` system so I studied how the book reading system works. I noticed (very noticeable) that there were quite a bit of redundant code/conditions and a lot of noise.

As a newbie, I was a little lost. Mostly because of the long conditionals and the bunch of strings everywhere. So I thought maybe I could do something about it. As such, most of what I'm gonna write below is from the perspective of a newbie contributor.

So for this PR, I focused on centralizing book readability checks and adding new helper data/functions.

#### Describe the solution

**1. Book mastery**
```
enum book_mastery {
    CANT_DETERMINE, // book not yet identified, so you don't know yet
    CANT_UNDERSTAND, // does not have enough skill to read
    LEARNING,
    MASTERED // can no longer increase skill by reading
};
book_mastery Character::get_book_mastery( book );
```
There were code blocks with this type of pattern.
```
const auto &type = it.type->book;
const skill_id &skill = type->skill;
const int lvl = get_skill_level( skill );
...
if( skill && lvl < type->req ) { ... }
else if ( skill && lvl < type->level ) { ... }
...
```
With the this util, we can then change it like so:
```
book_mastery mastery = get_book_mastery( it );
...
if ( mastery == book_mastery::CANT_UNDERSTAND ) { ... }
else if ( mastery == book_mastery::LEARNING ) { ... }
```
A simple example but as you can see, this new version has less symbols but more english. I'd say this is more readable and thus is an improvement. This function should also prove useful for writing unit tests.

**2. An enum for the different reasons why someone can't read a book**
```
read_fail_reason {
    NO_REASON,
    NOT_A_BOOK,
    DRIVING,
    LOW_MORALE,
    LOW_SKILL,
    ILLITERATE,
    NEED_READING_GLASSES,
    TOO_DARK,
    BLIND
};
```
Only generic reasons are included since all of them can be safely applied to both NPCs and the player. There are other NPC-related reasons but they made generalizing the code changes below more complicated.

**3. An all-in-one function for book readability checks**
```
struct readability_eval {
    bool can_read;
    bool can_be_assisted;
    bool can_learn;
    bool can_have_fun;
    read_fail_reason fail_reason;
    book_mastery mastery;
};
readability_eval Character::evaluate_readability( book );
```
This should be self explanatory but this helps answer the following questions in one go.

- Can this character read this? If not, why?
- Can others read for this character?
- Can this character learn from it?
- Can this character gain fun from it?

My main target for this was the `avatar::read` function which uses all of these fields. However, due to lack of tests, I decided I should do it for another PR cause I don't want to break something. It's still useful in terms of centralizing book readability checks. The player, NPCs reading alone, and NPCs reading for the player, all had different checks. So now they just use this function.

**4. Centralized generic read failure messages**
```
std::string Character::get_read_fail_message( read_fail_reason, book, is_alone );
```
As a bonus to the changes above, it was easy to centralize this. This takes care of giving the correct reason string based on its arguments which should help remove some noise. Only generic read failure messages are included for the same reason in #**2.**

**5. Misc changes**

I moved `fun_to_read()` and `book_fun_for()` from player into Character. I also change `item::get_remaining_chapters()` to now take a Character as parameter, instead of player.

**All-in-all**, most of these changes are just prep work. I'm *planning* to expand on these changes.

#### Describe alternatives you've considered

I was thinking of implementing the utility functions as free functions, perhaps in their own namespace. I decided against it since it seemed like most of the code is tied to classes (from what I have seen).

The utility functions could also be virtual and just let the avatar and npc classes handle their own thing. I thought this was unnecessary since the functions are generic/short enough.

#### Testing

All the tests passed. ~~As of submitting this PR, I'll be doing *some* playtesting. *Some* because testing group reading activities with multiple NPCs is quite tedious. I hope that's fine. Until then, I'll mark this PR WIP.~~

I tested reading alone and asking an NPC to read a book. I'm confident that both are working fine.

Now, in terms of reading with different NPCs, it was quite a challenge. Though I *believe* this also works from my playtesting. I tested reading with 1 other NPC more thoroughly than with more. My reason being if it works with 1 NPC, it should theoretically work for more. I tried different combinations of the following, both applying to the player and NPCs:

- illiterate trait
- far sighted trait with or without glasses
- blindfolds and the debug blind effect
- playing with skill level requirements

I thought about writing unit tests for this but I don't really see how given how the code is structured so... yeah.

**tldr;** Reading books and not being able to read books should work as expected. It's the reasons shown on screen for why you can't read which I'm not completely sure of.

#### Additional context

I only made 2 small contributions so far so I hope this is not too pretentious of me.

I mainly did this as an exercise so any criticisms/comments are welcome.

### CODE REVIEW

**Comment by sacrificerXY** [comment]
So far, in terms of actual gameplay effects, things seems to work as expected. You can and cannot read correctly. It's just that the reasons for why you can't read seems off. I think the culprit is the fact that NPCs are not required to identify books, which I forgot. I need to investigate this some more later.

**Comment by sacrificerXY** [comment]
I'll close this and I'll split up the basic refactor stuff to another PR.

Regarding the centralization idea, I think I'll scrap it for now and revisit it later. As more time passes, the more I realize how complicated the problem actually is. These were the things which made it tricky:

- NPCs doesn't need to identify books
- NPCs are exempted from some conditions
- NPCs and player have different condition priorities
- reading alone or in a group also changes the conditions
- only 1 failure condition matters, even if multiple conditions could fail in the following conditions

All these made it hard to debug why exactly something doesn't work (had to use the classic `std::cout` spam).

My PR works for the most part. I just feel like there is a better way to do it.
